Repository locking

dbnicholson opened the pull request:
In #813, some simple repository locking was being considered to prevent prunes from running concurrently with checkouts and transactions. While I think that would be a great thing, I don't think that level of locking is rich enough for ostree. Since the locking isn't recursive, it would prevent both libostree and ostree programs from obtaining an exclusive lock if it would also happen somewhere else in libostree. The simple example I can think of is `ostree_repo_prune_static_deltas()`. You'd want to take an exclusive lock there as an entry point to libostree, but it's also called from `ostree_repo_prune()`, which you'd also want to take an exclusive lock for. That would simply deadlock.

The alternative I came up with here is to maintain the lock in thread local storage with a stack of lock states. Then each thread locks independently and the stack stays coherent. ddfa34b125365b4a0dd293c95c68e4fb2bc1c2e7 has that implementation with a large comment in `ostree-repo.c` describing how it works. The API is then a push/pop interface.

The drawback of having the lock in TLS is that it seems to prevent taking the lock asynchronously since that would likely involve having another thread acquire the lock. Once that thread completed, the lock would be dropped. So, the implementation here blocks synchronously when pushing or popping the lock. The lock timeout can be managed globally with a repo config option or from an API.

All the new APIs are experimental, but I think it would be best to make them public so ostree programs can do their own locking around a series of ostree operations.

I added a bunch of locking in places I thought it would be needed. One spot I didn't try to tackle was the core and remote config. Thinking about all the use cases got me confused, so I punted for now...

This closes https://bugzilla.gnome.org/show_bug.cgi?id=759442.

Let me know what you think! I know there's a lot here, but this issue has affected us quite a bit at Endless.

cgwalters commented:
High level looks sane at a brief glance, but there's lots of details here.  (One I have is how this intersects with worker threads we create internally)

How about we avoid lots of `#ifdefs` here by having a private version that's used internally, then the "experimental" bit is just exposing it publicly?   The private version could actually be a repo setting too so it doesn't depend on the experimental-API build?

dbnicholson commented:
> High level looks sane at a brief glance, but there's lots of details here. (One I have is how this intersects with worker threads we create internally)

Yeah, I tried to think about that a lot, but I couldn't totally come to grips with it. Basically, the worker threads are OK if you only ever acquire a shared lock. Which (I believe) is the case now. Otherwise, you have to arrange to take the lock before the worker is spawned and ensure that the worker doesn't go through a path that acquires locks.

I spent a lot of time thinking about that how to address that. I had another version that tried to detect if the lock was being acquired in the same thread by stashing the main context in use when the repo was opened and skipping if the main context in use when the lock was acquired wasn't the same. But that only tells you if the context is the same, not if the thread is the same. (I might be missing something there. I am definitely not the `GMainContext` guru.)

The only other thing I could think of was having a dedicated locking thread, but I then didn't know how to address the lock stack being updated out of order by multiple threads.

> How about we avoid lots of #ifdefs here by having a private version that's used internally, then the "experimental" bit is just exposing it publicly?

You mean in `src/ostree`? I could add a `cmdprivate` pointer if that's better. Most of `src/libostree` is clean except for the headers.

> The private version could actually be a repo setting too so it doesn't depend on the experimental-API build?

Not sure I follow here. What would the setting be?

cgwalters commented:
Yeah, I mean the binaries would use a cmdprivate API to do locking, which we could then enable by setting e.g.
```
[core]
locking=true
```
or so in the repo config.  (Which itself we could label as experimental I guess).   That'd allow us to land most of these patches and gain experience with it a bit, even on builds that don't `--enable-experimental-api`.

dbnicholson commented:
Oh, well I only added actual locking in `fsck`. The rest of the stuff in `src/ostree` is just turning the `--lock-timeout` option into a call to `ostree_repo_set_lock_timeout()`, which could be `cmdprivate` like you say.

However, the repo config option could be used to make `ostree_repo_lock_push()` and `ostree_repo_lock_pop()` into noops. That's a good idea in case this is broken...

I also see that the CI failed immediately because I haven't actually tested without `--enable-experimental-api` in a while. Whoops.

pwithnall commented on man/ostree.repo-config.xml:
```diff
@@ -143,6 +143,18 @@ Boston, MA 02111-1307, USA.
+        <listitem>
+          <para>
+            Integer value controlling the default repository locking
+            timeout in seconds. If the value is <literal>-1</literal>,
```
What if the value is 0 or <-1? Could go with “If the value is negative then ostree will block…”

dbnicholson replied:
0 means don't wait - true non-blocking. I thought that was implied, but I could spell it out. < -1 is an error. I could just do all negative values, but I liked having the explicit value. It also allowed me to later use -2 as a sentinel.

pwithnall commented on man/ostree.xml:
```diff
@@ -120,6 +120,23 @@ Boston, MA 02111-1307, USA.
+                  Some <command>ostree</command> commands require the
+                  repository to be locked while operating. This option sets
+                  the repository lock timeout to TIMEOUT seconds. If TIMEOUT
+                  is <literal>-1</literal>, then the locking will block
```
Similarly here: what if it’s <-1 or 0?

dbnicholson replied:
OK.

pwithnall commented on src/libostree/ostree-repo-private.h:
```diff
@@ -413,8 +418,36 @@ _ostree_repo_get_remote_inherited (OstreeRepo  *self,
+
+typedef OstreeRepo OstreeRepoAutoLock;
+OstreeRepoAutoLock * ostree_repo_auto_lock_push (OstreeRepo    *self,
+                                                 gboolean       exclusive,
```
Might make sense to use an enum or flags instead of a boolean here, since it’s typically not obvious what the sense of a boolean is. Does `TRUE` mean ‘make it exclusive’ or ‘make it shared’?

dbnicholson replied:
Yeah, I thought about that later after doing `TRUE`/`FALSE` so many times. It seemed sort of silly to have an enum that would only ever have 2 exclusive states. I guess:
```
typedef enum {
  OSTREE_REPO_LOCK_SHARED,
  OSTREE_REPO_LOCK_EXCLUSIVE
} OstreeRepoLockType;
```

pwithnall replied:
I think people (me included) should get out of the habit of thinking of two-element enums as ‘silly’, if they improve code readability. That said, there’s the possibility that this could be extended in future to include more flags or enum elements for more esoteric extensions to the locking.

dbnicholson replied:
OK, I'll do it. That said, I would like this to remain a simple enum and not migrate into some flags setup later.

pwithnall commented on src/libostree/ostree-repo-private.h:
```diff
@@ -37,6 +37,9 @@ G_BEGIN_DECLS
 #define _OSTREE_MAX_OUTSTANDING_FETCHER_REQUESTS 8
 #define _OSTREE_MAX_OUTSTANDING_DELTAPART_REQUESTS 2
 
+/* Keep this in sync with the man page for the core.lock-timeout option. */
```
Mention that it’s in seconds.

dbnicholson replied:
Sure.

pwithnall commented on src/libostree/ostree-repo-private.h:
```diff
@@ -153,6 +157,7 @@ struct OstreeRepo {
   guint64 tmp_expiry_seconds;
   gchar *collection_id;
   gboolean add_remotes_config_dir; /* Add new remotes in remotes.d dir */
+  gint lock_timeout;
```
Maybe name it `lock_timeout_seconds` to make the units more obvious? Same for the `#define` above.

dbnicholson replied:
Sure

pwithnall commented on src/libostree/ostree-repo-prune.c:
```diff
@@ -401,6 +417,11 @@ ostree_repo_prune_from_reachable (OstreeRepo        *self,
+  g_autoptr(OstreeRepoAutoLock) lock = NULL;
+  lock = ostree_repo_auto_lock_push (self, TRUE, cancellable, error);
+  if (!lock)
+    return FALSE;
```
One thought: if you changed the definition of `ostree_repo_auto_lock_push()` to:
`gboolean ostree_repo_auto_lock_push (OstreeRepo *self, OstreeRepoLockFlags flags, OstreeRepoAutoLock **out_lock, GCancellable *cancellable, GError **error)`, then all the uses of it could simplify to:
```
g_autoptr(OstreeRepoAutoLock) lock = NULL;
if (!ostree_repo_auto_lock_push (self, OSTREE_REPO_LOCK_EXCLUSIVE, &lock, cancellable, error))
  return FALSE;
```

dbnicholson replied:
I could go either way on this and did consider that interface. I do generally like ostree's boolean return values with supplemental information in out params. In this case it seemed overkill to me, though, as both the return value and the out param would be transmitting the same information. Would the out param be a requirement? it would be useless to use the function otherwise. Not that we're counting cycles in ostree, but it's also less efficient to put another parameter on the stack and fill it in.

You can, of course, do the call and check in one line if you want with the current interface:
```
if (!(lock = ostree_repo_auto_lock_push(...)))
  return FALSE;
```
That's definitely not as aesthetically pleasing as the interface you've suggested, which is also why I prefer to write it on 2 lines.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,508 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+          cur_state_name = "null";
+        }
+
+      g_debug("Free lock: state=%s, depth=%u", cur_state_name, stack_len);
```
Nitpick: Missing space before `(` (and in a couple of other places).

dbnicholson replied:
Good catch.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -5099,3 +5693,47 @@ ostree_repo_set_collection_id (OstreeRepo   *self,
+ * configured in the core.lock-timeout option will replace this.
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ * Since: 2017.8
```
Since: 2017.13

(probably wrong in a few other places)

dbnicholson replied:
Bah. I actually think I was pretty good about setting it to 2017.13, but I'll skim through for that. Thanks for catching that.

pwithnall commented on src/libostree/ostree-repo.h:
```diff
@@ -108,6 +108,40 @@ OstreeRepo *  ostree_repo_create_at (int             dfd,
+/**
+ * OstreeRepoAutoLock: (skip)
+ *
+ * This is simply an alias tox #OstreeRepo used for automatic lock cleanup.
```
s/tox/to/

dbnicholson replied:
Thanks, I saw that myself later.

cgwalters commented:
Yeah...we should probably try to move fsck into the library too.  But doesn't need to block this patch.

Let's try to get this cleaned up and then I'm happy to merge it and we can do followup commits?

dbnicholson commented:
@cgwalters Had another thought about the threads.

* If the lock to be taken is shared, only do it when outside a transaction. Assuming that the transaction is started on a thread that will live throughout the life of the transaction, then you already have a shared lock.

* If the lock to be taken is exclusive, add internal _unlocked versions that would be called in any place that would be run from a worker thread. So far I've only encountered `ostree_repo_delete_object` removing tombstones when writing out commits.

I have a bunch of fixups that I'll push out shortly that should address the comments so far.

dbnicholson commented:
OK, I added a bunch of fixups that I'll squash soon, but I wanted to give a chance to see the intermediate state if anyone was interested. I believe I addressed most of the comments so far.

TODO:

* Add the `core.locking` option to globally enable or disable locking
* Investigate locking from worker threads more
* Look at locking for core and remote config

cgwalters commented:
Are you going to have a chance to rebase this soon?  If not I might be able to do it early this week.

dbnicholson commented:
Yeah, I'll rebase later today. Some of the commits should be reordered a bit.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+
+/* Wrapper to handle flock vs OFD locking based on GLnxLockFile */
+static gboolean
+do_repo_lock(int fd,
```
Nitpick: Missing space before `(`. Same for various methods below.

dbnicholson replied:
Blah. I thought I was good about that.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+  if (res < 0)
+    {
+      if (errno == EINVAL)
+        res = flock (fd, flags);
```
This needs a `TEMP_FAILURE_RETRY` loop.

dbnicholson replied:
Yeah, this can also get `EINTR` when you're doing a blocking lock.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+#ifdef F_OFD_SETLK
+  struct flock fl = {
+    .l_type = (flags & ~LOCK_NB) == LOCK_EX ? F_WRLCK : F_RDLCK,
+    .l_whence = SEEK_SET,
```
Might be clearer to explicitly initialise `.l_start` and `.l_len` to 0 here.

dbnicholson replied:
I think you're right. I copied this from `GLnxLockFile`, but I agree that explicitly setting the members is clearer.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+    .l_whence = SEEK_SET,
+  };
+
+  res = fcntl (fd, (flags & LOCK_NB) ? F_OFD_SETLK : F_OFD_SETLKW, &fl);
```
Needs a `TEMP_FAILURE_RETRY` loop.

dbnicholson replied:
Good call.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+#ifdef F_OFD_SETLK
+  struct flock fl = {
+    .l_type = F_UNLCK,
+    .l_whence = SEEK_SET,
```
Same comments as above function about explicitly initialising `.l_start` and `.l_len`, and `TEMP_FAILURE_RETRY`.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
+  if (lock == NULL)
+    {
+      lock = g_new (OstreeRepoLock, 1);
```
Nitpick: I’d always use `g_new0()` for paranoia reasons.

dbnicholson replied:
Heh. I actually did that intentionally for similar paranoid reasons. If `free_repo_lock` is called and `lock->fd == 0`, then you'd close `stdin`. So, I explicitly ordered the initialization below so you wouldn't do that. But I guess the compiler could do something different either way... I'll change it to `g_new0` so we don't accidentally free random memory.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+      lock = g_new (OstreeRepoLock, 1);
+      lock->stack = g_queue_new ();
+      g_debug ("Opening repo lock file");
+      lock->fd = openat (self->repo_dir_fd, ".lock",
```
I think this also needs `TEMP_FAILURE_RETRY`.

dbnicholson replied:
Good call.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+  else
+    {
+      cur_state = GPOINTER_TO_INT (g_queue_peek_head (lock->stack));
+      cur_state_name = (cur_state == LOCK_EX) ? "exclusive" : "shared";
```
The code for calculating `cur_state_name` and `cur_state` is present in a couple of functions and could be abstracted out.

dbnicholson replied:
I suppose. It's so simple, but I guess that would be nicer.

dbnicholson replied:
Fixed in #1343. I broke out a little helper and struct that seems to work correctly.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+ * Takes a lock on the repository and adds it to the lock stack. If @lock_type
+ * is %OSTREE_REPO_LOCK_SHARED, a shared lock is taken. If @lock_type is
+ * %OSTREE_REPO_LOCK_EXCLUSIVE, an exclusive lock is taken. shared lock is
+ * used. The actual lock state is only changed when locking a previously
```
Something looks wrong with the ‘shared lock is used’ mini-sentence. What’s it supposed to mean?

dbnicholson replied:
Whoops. Probably a rebase mishap. I'll clean that up.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+                       GError             **error)
+{
+  g_return_val_if_fail (self != NULL, FALSE);
+  g_return_val_if_fail (self->inited, FALSE);
```
Missing preconditions:
```
g_return_val_if_fail (OSTREE_IS_REPO (self), FALSE);
g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable), FALSE);
g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
```

dbnicholson replied:
Sure. I think you could make that argument for most every API in ostree, though...

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+      g_debug ("Pushing lock blocking");
+      return push_repo_lock (self, lock_type, TRUE, error);
+    }
+  else
```
Maybe add a `g_assert (self->lock_timeout_seconds >= 0)` here.

dbnicholson replied:
Since any setter for `self->lock_timeout_seconds` already checks that it's `>= -1`, I decided not to bother here. But I could add the assert to make sure we don't mess up somewhere.

pwithnall replied:
I find such assertions help to document the code, and also help keep static analysis tools on the straight and narrow.

dbnicholson replied:
Fixed in #1343. I put an assert for `>= -1` above the entire block.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+    }
+  else
+    {
+      gint waited = 0;
```
`waited` could be unsigned, and you could convert the non-negative `lock_timeout_seconds` to unsigned within this block.

dbnicholson replied:
What's the advantage of that?

pwithnall replied:
Eliminates the question of ‘what if `waited` ends up being negative?’.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+      gint waited = 0;
+      g_debug ("Pushing lock non-blocking with timeout %d",
+               self->lock_timeout_seconds);
+      for (;;)
```
`while (TRUE)` might make it a bit more explicit that this loop is potentially infinite.

dbnicholson replied:
Doesn't make a big difference to me. `for (;;)` is used elsewhere in the ostree code, so I used that. Both constructs imply potentially infinite looping to me.

pwithnall replied:
*shrug* :-)

dbnicholson replied:
Stlll sticking with `for(;;)` unless there's a tie-breaker :)

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+}
+
+/**
+ * ostree_repo_lock_pop
```
Missing a `:` from the end of the function name (same problem in other gtk-doc comments).

dbnicholson replied:
Right. I'd seen it there sometimes and sometimes not within ostree, but I agree that's the canonical format.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+                      GError       **error)
+{
+  g_return_val_if_fail (self != NULL, FALSE);
+  g_return_val_if_fail (self->inited, FALSE);
```
Preconditions missing as in previous function.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+      g_debug ("Popping lock blocking");
+      return pop_repo_lock (self, TRUE, error);
+    }
+  else
```
I suggest adding a `g_assert()` here and converting `lock_timeout_seconds` to unsigned within this block, as in the previous function.

dbnicholson replied:
Fixed in #1343. As before, added an assert for `>= -1` above both blocks.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+      gint waited = 0;
+      g_debug ("Popping lock non-blocking with timeout %d",
+               self->lock_timeout_seconds);
+      for (;;)
```
Suggest `while (TRUE)` as in the previous function.

dbnicholson replied:
Sticking with `for(;;)` as above.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -2271,6 +2783,38 @@ reload_core_config (OstreeRepo          *self,
+        gint64 lock_timeout = g_ascii_strtoll (lock_timeout_str, &endptr, 0);
+        if ((lock_timeout < -1) ||
+            (lock_timeout > G_MAXINT) ||
+            (lock_timeout == 0 && endptr == lock_timeout_str))
```
This doesn’t check for the case where `*endptr != '\0'`, which could be caused by `lock_timeout_str == "123somestring"`.

dbnicholson replied:
Good catch. I hadn't thought of that.

pwithnall replied:
The `g_ascii_strto*()` are an absolute nightmare for correct error handling.

GLib recently acquired `g_ascii_string_to_{un,}signed()`, which are a lot better, but unfortunately too recent for libostree to use yet. :-(

dbnicholson replied:
Will fix after #1343. This part didn't make it into the slimmed down setup.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -5099,3 +5702,47 @@ ostree_repo_set_collection_id (OstreeRepo   *self,
+ * represents the number of seconds the application will sleep attempting to
+ * acquire the repository lock.
+ *
+ * Returns: lock timeout for the repository
```
Nitpick: maybe add `, in seconds` to this?

dbnicholson replied:
Sure.

dbnicholson replied:
Will fix after #1343.

pwithnall commented on src/libostree/ostree-repo.c:
```diff
@@ -5099,3 +5702,47 @@ ostree_repo_set_collection_id (OstreeRepo   *self,
+ * Set the lock timeout for the repository. A timeout of -1 indicates that the
+ * application will block until the lock is acquired. Otherwise, it represents
+ * the number of seconds the application will sleep attempting to acquire the
+ * repository lock.
```
Should mention that anything `< -1` is not allowed.

dbnicholson replied:
OK

dbnicholson replied:
Will fix after #1343.

pwithnall commented on src/libostree/ostree-repo.h:
```diff
@@ -107,6 +107,51 @@ OstreeRepo *  ostree_repo_create_at (int             dfd,
+ * OstreeRepoLockType:
+ * The type of repository lock to acquire.
+ * @OSTREE_REPO_LOCK_SHARED: A shared lock
+ * @OSTREE_REPO_LOCK_EXCLUSIVE: An exclusive lock
```
Missing a `Since:` line.

dbnicholson replied:
Good catch.

dbnicholson replied:
Fixed in #1343.

pwithnall commented on src/ostree/ot-main.c:
```diff
@@ -48,6 +50,28 @@ static GOptionEntry repo_entry[] = {
+  gint64 lock_timeout = g_ascii_strtoll (value, &endptr, 0);
+  if ((lock_timeout < -1) ||
+      (lock_timeout > G_MAXINT) ||
+      (lock_timeout == 0 && endptr == value))
```
As with the other `g_ascii_strtoll()` call, this should check for `*endptr != '\0'` as another error condition.

dbnicholson replied:
Will fix after #1343.

dbnicholson commented on Makefile-libostree.am:
```diff
@@ -261,6 +261,10 @@ OSTree-1.0.gir: libostree-1.la Makefile
 OSTree_1_0_gir_INCLUDES = Gio-2.0
 OSTree_1_0_gir_CFLAGS = $(libostree_1_la_CFLAGS)
+if ENABLE_EXPERIMENTAL_API
+# When compiling this is set via config.h, but g-ir-scanner doesn't use that
```
Interesting, I didn't know that. I'll try that locally.

dbnicholson replied:
This was separately fixed in #1322.

dbnicholson commented on src/libostree/ostree-cmdprivate.c:
```diff
@@ -1,3 +1,4 @@
+
```
Whoops...

dbnicholson replied:
Will address after #1343.

dbnicholson commented on src/libostree/ostree-cmdprivate.c:
```diff
@@ -48,7 +49,14 @@ ostree_cmd__private__ (void)
     _ostree_repo_static_delta_query_exists,
-    _ostree_repo_static_delta_delete
+    _ostree_repo_static_delta_delete,
+    /* Remove this when ostree_repo_set_lock_timeout is no longer experimental */
```
Sure

dbnicholson replied:
Will address after #1343.

dbnicholson commented on src/libostree/ostree-cmdprivate.h:
```diff
@@ -20,17 +20,34 @@
 #pragma once
 
 #include "ostree-types.h"
+/* This is only needed to get the OstreeRepoLockType and OstreeRepoAutoLock
```
Sure

dbnicholson replied:
Will address after #1343.

dbnicholson commented on src/libostree/ostree-repo-checkout.c:
```diff
@@ -991,6 +992,16 @@ checkout_tree_at (OstreeRepo                        *self,
     }
 
+  /* Take a shared repo lock to try to ensure objects aren't deleted. If the
+   * repo is not writable, this will be a noop and we just hope for the
```
The case is simply an unprivileged user trying to checkout from /ostree/repo or /var/lib/flatpak/repo or something like that. They should be allowed to do a checkout as they were before.

Like you say, if the repo is not writable by the current user, then they're not going to be deleting any objects. But if, say, flatpak is pruning /var/lib/flatpak/repo while you're trying to do a checkout, bad things could happen.

But that's already the case, and I don't think we should optimize for unprivileged users. I'm really just trying to make sure that multiple processes running as the repo owner can't screw each other since that's by far the common case. Since the unprivileged user won't even be able to open the lock file (it's opened 600 particularly so that only repo owners can take exclusive locks), then I don't think they'll be able to get the lock state.

An inotify watch is interesting. I'm not sure inotify can tell you when the lock state changes, though, only if it's been opened.

pwithnall replied:
Yeah, having done a bit of reading, I don’t think it’s possible to monitor the status of a file lock using inotify. The only way of doing it is to poll with `fcntl(F_GETLK)`, but that’s horrific and racy.

The other approach I can think of is to use a world-writeable directory (somewhere in `/run`, probably), which can contain the locks for many repositories, each with a deterministic name based on the repository directory’s device ID and inode.

That said, allowing an unprivileged process to take a lock on a repository it can’t write to allows priority inversion attacks, since there’s no realistic way to stop it taking an exclusive lock.

Unless we put shared locks in `/run` and exclusive locks in `$repo/.lock`?

dbnicholson replied:
I still believe the way this was written where only the repo owner can manage the lock is correct. I think any attempt to allow non-owners to take a lock would be very complex and likely mean that flock-type locking can't be used.

dbnicholson commented on src/libostree/ostree-repo.c:
```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+ * repository.
+ *
+ * The intended usage is to take a shared lock when writing objects or
+ * reading objects in critical sections. Exclusive locks are taken when
```
Yeah, it's something I punted on because the semantics are a little weird. Internally, ostree reads it into memory once and then it's accessed via `ostree_repo_get_config` or `ostree_repo_copy_config`. I don't think locking there would be helpful.

What I think you could do is take a shared lock during `ostree_repo_reload_config` (when the files are actually read) and take an exclusive lock during `ostree_repo_write_config` and `ostree_repo_remote_{add,change,delete}`. But then I think it can't really be managed all the way internally because the typical pattern is `ostree_repo_copy_config`, update keyfile, `ostree_repo_write_config`. You'd really want to take an exclusive lock around the whole set so someone else doesn't change it while you're making keyfile modifications. That's certainly something flatpak does. But I didn't think all the way through it so punted for now.

dbnicholson replied:
Oh, and `ostree_repo_reload_config` is called during `ostree_repo_open`, so I'd be worried about locking before you could even open the repo. But I guess you could probably skip the locking there if `self->inited` is `FALSE`.

pwithnall replied:
Seems reasonable to me to punt on this for now, but please file a follow-up issue about it.

dbnicholson replied:
Will address after #1343.

cgwalters commented:
Would be happy to start trying to take this - perhaps it's simplest to split this up into a PR that adds the internal locking API with just a few uses - maybe just prune and commit?  Then we can gradually trickle in changes to enable it in different places like fsck as we test?

dbnicholson commented:
Sorry, got busy with a bunch of other stuff but I should be getting back to this today or tomorrow. I'll split it up into 2 PRs like you say. Some other stuff I've thought of in the meantime:

* A repo wide locking enable boolean. I hope this wouldn't be necessary much, but it would allow programs to opt out of locking if we run into issues with this or if there's a sequence that's just too complex for the internal locking.

* A thread specific locking enable boolean. This is my idea for handling the worker threads. What you'd do is take a lock, spawn the thread, and immediately disable locking for that thread. When the thread completed, the spawner would pop the lock. That would allow (for example) all the pull threads to run without locking while the thread spawner has the repo locked and avoid the issues with

dbnicholson commented:
I opened #1343 with the slimmed down initial setup as requested. I still passes the test suite, although I haven't really used it in real applications.

cgwalters commented:
How do you want to proceed with rebasing this?  Are you planning to work on it anytime soon?

dbnicholson commented:
Sorry, I had to take off my ostree developer hat for a while. Next week I'll get back on this and flesh it out. Does that timeline work or were you looking for something sooner?

cgwalters commented:
No pressure on the timeline!  Currently locking is disabled by default.  I am happy to help out a bit if we can figure out how to subdivide work.  Just trying to keep things moving!
